### PR TITLE
Executor2

### PR DESCRIPTION
**Pull request by wotyas**
As previously requested, we want to merge dev branch with recently added IO Modules.
Feel free to comment and contribute in code.

### CODE REVIEW

**Comment by wotyas** [comment]
Added requested files in last commit.

File: README.md [hunk 1]
```diff
@@ -1,49 +1,49 @@
...
+ -
+ -## Contribution
+ -* Create branch describing what are you working on. For example Executor, Loaders, Plugins
+ -* Commit your changes often [Why?](http://www.databasically.com/2011/03/14/git-commit-early-commit-often/)
+ -* When you finish small part of your module create pull request. [How?](https://help.github.com/articles/creating-a-pull-request/)
+ -* Pull request has to be small <200 changes. If you will make big pull request be aware that it might take very long time to merge it and you will be slowed down
+ -. [Why?](http://robertheaton.com/2015/10/26/why-and-how-to-make-smaller-pull-requests/)
+ -* Schema for code pushing is:
+ -Your branch -> Pull Request -> dev -> Pull Request -> stage -> Pull Request -> master
+ -[Link](https://help.github.com/articles/about-pull-request-reviews/)
```
Comment (slaby93) [inline README.md]: Still Invalid - you added "-" signs everywhere

Comment (slaby93) [inline README.md]: also here

File: SampleUsage/src/LoadEveryFilesInDirAndProcess.java [hunk 2]
```diff
@@ -1,16 +0,0 @@
...
-        e.loadFromDir("path/to/dir/with/files");
-
-        e.addPlugin(JImageStreamToolkit.plugins.Grayscale());
-
-        e.process();
-
-        ResultImagesSet result = e.getResultSet();
-
-    }
-}
```
Comment (slaby93) [inline SampleUsage/src/LoadEveryFilesInDirAndProcess.java]: why did You remove this file?

File: SampleUsage/src/LoadManyFilesAndProcess.java [hunk 3]
```diff
@@ -1,27 +0,0 @@
...
-
-        e.addPlugin(JImageStreamToolkit.plugins.Grayscale());
-
-        e.process();
-
-        ResultImagesSet result = e.getResultSet();
-
-
-    }
-}
```
Comment (slaby93) [inline SampleUsage/src/LoadManyFilesAndProcess.java]: Also this one?

File: src/IOModules/AllGraphicLoaders.java [hunk 4]
```diff
@@ -0,0 +1,38 @@
+package IOModules;
```
Comment (slaby93) [inline src/IOModules/AllGraphicLoaders.java]: steven image is added two times why?

File: src/IOModules/AllGraphicLoaders.java [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+ */
+
+public class AllGraphicLoaders {
+
+    private static Map<String, GraphicLoader> loaders = null;
+
+    public static Map<String, GraphicLoader> getGraphicLoaders(){
+
+        if(loaders != null){
+            return loaders;
```
Comment (slaby93) [inline src/IOModules/AllGraphicLoaders.java]: You do not need two returns here. If loaders are null then init them, else return also loaders initialization could be moved to another function to keep SOLID

File: src/IOModules/BMPLoader.java [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+    @Override
+    public Collection<Image> load(ImgFile img) throws IOException {
+
+        Collection<Image> images = new ArrayList<>();
+        BufferedImage tmpImage = ImageIO.read(new File(img.getPath()));
+        images.add(new BasicImage(tmpImage));
+
+        return images;
+    }
+}
```
Comment (slaby93) [inline src/IOModules/BMPLoader.java]: OFFTOP: at the end different file types contain a lot of meta data. Loaders like this could retrive those data and attach it as metadata object to image

File: src/IOModules/BasicImage.java [hunk 7]
```diff
@@ -0,0 +1,42 @@
+package IOModules;
+
+import javax.imageio.ImageIO;
+import java.awt.image.BufferedImage;
+import java.io.File;
+import java.io.IOException;
+
+public class BasicImage extends Image {
```
Comment (slaby93) [inline src/IOModules/BasicImage.java]: Add docs explaining what is basic image ( how it differs from other images? )

File: src/IOModules/BasicImage.java [hunk 8]
```diff
@@ -0,0 +1,42 @@
+package IOModules;
+
+import javax.imageio.ImageIO;
+import java.awt.image.BufferedImage;
+import java.io.File;
+import java.io.IOException;
+
+public class BasicImage extends Image {
+
+    private BufferedImage image;
```
Comment (slaby93) [inline src/IOModules/BasicImage.java]: Boss Misztal asked to look for another image format ( if possible )

File: src/IOModules/BasicImage.java [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+    }
+
+    @Override
+    public int getHeight() {
+        return image.getHeight();
+    }
+
+    @Override
+    public int getRGB(int x, int y) {
+        return image.getRGB(x, y);
```
Comment (slaby93) [inline src/IOModules/BasicImage.java]: what if we will get x,y out of image zone eg. 
x = -999999999
y = 999999999

?

File: src/IOModules/BasicImage.java [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+    }
+
+    @Override
+    public void setRGB(int x, int y, int value) {
+        image.setRGB(x, y, value);
+    }
```
Comment (slaby93) [inline src/IOModules/BasicImage.java]: what if I will pass value like 1238192 ? Also isn't it RGBA ?

File: src/IOModules/JPGLoader.java [hunk 11]
```diff
@@ -0,0 +1,26 @@
+package IOModules;
+
+import javax.imageio.ImageIO;
+import java.awt.image.BufferedImage;
+import java.io.File;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.io.IOException;
+
+public class JPGLoader extends GraphicLoader {
```
Comment (slaby93) [inline src/IOModules/JPGLoader.java]: JPG and JPEG loaders as basically same code. Loader should return array of possible extensions instead of just one String value ( loaders can load multiple types of files )

File: src/Jimage/src/Executor.java [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.Collection;
+import java.util.stream.Stream;
+
+public abstract class Executor {
```
Comment (slaby93) [inline src/Jimage/src/Executor.java]: what is this? Why dou You create executor Class?

File: src/Jimage/src/Params.java [hunk 13]
```diff
@@ -0,0 +1,4 @@
+package Jimage.src;
+
+public class Params {
+}
```
Comment (slaby93) [inline src/Jimage/src/Params.java]: ?????

File: src/Modules/ChainOfPhoto.java [hunk 14]
```diff
@@ -0,0 +1,19 @@
+package Modules;
```
Comment (slaby93) [inline src/Modules/ChainOfPhoto.java]: third stiven

File: src/Modules/ModPNGHandler.java [hunk 15]
```diff
@@ -0,0 +1,18 @@
...
+    @Override
+    void saveImage(BufferedImage image, FileExtension extension,String newFileName) throws IOException {
+        if(extension.toString().equals(FileExtension.png.toString())){
+            File outputfile = new File(newFileName +"."+extension);
+            ImageIO.write(image, extension.toString(), outputfile);
+        }else {
+            forward(image,extension,newFileName);
+        }
+    }
+}
```
Comment (slaby93) [inline src/Modules/ModPNGHandler.java]: and two more stivens -.-

File: src/IOModules/BasicImage.java (hunk 8 above)
Comment (agarak) [inline src/IOModules/BasicImage.java]: Plugins group needs ImageInputStream (we are using it to get pixels to transform it) and ImageOutputStream (use to sends pixels back). Class Pixel be use to store one single pixel on the image.

Comment (jgprogramming) [inline src/IOModules/BasicImage.java]: http://marvinproject.sourceforge.net/javadoc/marvin/image/MarvinImage.html - check this out, unpacked BufferedImage with as low overhead on set/get operations as possible (without using some kind of cache or smth).

File: src/IOModules/BasicImage.java (hunk 9 above)
Comment (jgprogramming) [inline src/IOModules/BasicImage.java]: bufferedImage throws outOfBoundExceptions, handling it outside of getRGB method could be better

File: src/IOModules/BasicImage.java (hunk 10 above)
Comment (jgprogramming) [inline src/IOModules/BasicImage.java]: Not every image has alpha channel, you could convert it to one standard (https://github.com/kmisztal/JImageStream/blob/master/jis-api/src/main/java/pl/edu/uj/JImageStream/utils/ImageUtils.java). Be wary of premultiplied alpha, and saving images with alpha to jpg

File: src/IOModules/AllGraphicLoaders.java [hunk 16]
```diff
@@ -1,38 +1,41 @@
...
 
     public static Map<String, GraphicLoader> getGraphicLoaders(){
-
-        if(loaders != null){
-            return loaders;
+        if( loaders == null ){
+            init();
         }
+        return loaders;
+    }
```
Comment (slaby93) [inline src/IOModules/AllGraphicLoaders.java]: perfect <3 <3 <3
